# Segmenting and Clustering Neighborhoods in Toronto

## Part 1 - downloading data and pre-processing

For the Toronto neighborhood data, a Wikipedia page exists that has all the information we need to explore and cluster the neighborhoods in Toronto. You will be required to scrape the Wikipedia page and wrangle the data, clean it, and then read it into a pandas dataframe so that it is in a structured format like the New York dataset.
https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

Built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name.

In [1]:
!wget -q -O 'neighborhoods.html' 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

In [2]:
from bs4 import BeautifulSoup
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [3]:
with open("neighborhoods.html") as filehandle:
    soup = BeautifulSoup(filehandle)

Find all tables in the HTML file

In [4]:
tables = soup.find_all(["table"])

We can see that the first table contains desired data

In [5]:
#print(tables[0])

Parse data from the table
- find all rows in the table
- first row contains headers
- remaining rows contain values

In [6]:
postal_codes = []
boroughs = []
neighborhoods = []
for row in tables[0].find_all(["tr"]):
    n = 1
    for col in row.find_all(["td", "th"]):
        data = col.string.strip()
        if n == 1:
            postal_codes.append(data)
        elif n==2:
            boroughs.append(data)
        else:
            neighborhoods.append(data)
        n = n + 1
print(postal_codes)

['Postal Code', 'M1A', 'M2A', 'M3A', 'M4A', 'M5A', 'M6A', 'M7A', 'M8A', 'M9A', 'M1B', 'M2B', 'M3B', 'M4B', 'M5B', 'M6B', 'M7B', 'M8B', 'M9B', 'M1C', 'M2C', 'M3C', 'M4C', 'M5C', 'M6C', 'M7C', 'M8C', 'M9C', 'M1E', 'M2E', 'M3E', 'M4E', 'M5E', 'M6E', 'M7E', 'M8E', 'M9E', 'M1G', 'M2G', 'M3G', 'M4G', 'M5G', 'M6G', 'M7G', 'M8G', 'M9G', 'M1H', 'M2H', 'M3H', 'M4H', 'M5H', 'M6H', 'M7H', 'M8H', 'M9H', 'M1J', 'M2J', 'M3J', 'M4J', 'M5J', 'M6J', 'M7J', 'M8J', 'M9J', 'M1K', 'M2K', 'M3K', 'M4K', 'M5K', 'M6K', 'M7K', 'M8K', 'M9K', 'M1L', 'M2L', 'M3L', 'M4L', 'M5L', 'M6L', 'M7L', 'M8L', 'M9L', 'M1M', 'M2M', 'M3M', 'M4M', 'M5M', 'M6M', 'M7M', 'M8M', 'M9M', 'M1N', 'M2N', 'M3N', 'M4N', 'M5N', 'M6N', 'M7N', 'M8N', 'M9N', 'M1P', 'M2P', 'M3P', 'M4P', 'M5P', 'M6P', 'M7P', 'M8P', 'M9P', 'M1R', 'M2R', 'M3R', 'M4R', 'M5R', 'M6R', 'M7R', 'M8R', 'M9R', 'M1S', 'M2S', 'M3S', 'M4S', 'M5S', 'M6S', 'M7S', 'M8S', 'M9S', 'M1T', 'M2T', 'M3T', 'M4T', 'M5T', 'M6T', 'M7T', 'M8T', 'M9T', 'M1V', 'M2V', 'M3V', 'M4V', 'M5V', 'M6V

Create pnadas data frame

In [7]:
toronto_df = pd.DataFrame({"Postal Code":postal_codes[1:], "Borough": boroughs[1:], "Neighborhood": neighborhoods[1:]})
toronto_df.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned                           
1         M2A      Not assigned                           
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Eliminate records with Not assigned Borough

In [8]:
toronto_df = toronto_df[toronto_df["Borough"]!='Not assigned']
toronto_df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
8           M9A         Etobicoke   
9           M1B       Scarborough   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
17          M9B         Etobicoke   
18          M1C       Scarborough   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
26          M9C         Etobicoke   
27          M1E       Scarborough   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
36          M1G       Scarborough   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N              York   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
107         M9P         Etobicoke   
108         M1R       Scarborough   
109         M2R        North York   
111         M4R   Central Toronto   
112         M5R   Central Toronto   
113         M6R      West Toronto   
114         M7R       Mississauga   
116         M9R         Etobicoke   
117         M1S       Scarborough   
120         M4S   Central Toronto   
121         M5S  Downtown Toronto   
122         M6S      West Toronto   
126         M1T       Scarborough   
129         M4T   Central Toronto   
130         M5T  Downtown Toronto   
135         M1V       Scarborough   
138         M4V   Central Toronto   
139         M5V  Downtown Toronto   
142         M8V         Etobicoke   
143         M9V         Etobicoke   
144         M1W       Scarborough   
147         M4W  Downtown Toronto   
148         M5W  Downtown Toronto   
151         M8W         Etobicoke   
152         M9W         Etobicoke   
153         M1X       Scarborough   
156         M4X  Downtown Toronto   
157         M5X  Downtown Toronto   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Vic

### Trying to find - More than one neighborhood that can exist in one postal code area.

In [9]:
pc = toronto_df[["Postal Code"]]
bg= pc[pc.duplicated()]
bg.head()

Empty DataFrame
Columns: [Postal Code]
Index: []

**Result:** There are no duplicated Postal Codes, so there are no areas assigned to two postal codes.

### If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

In [10]:
no_neighborhood = toronto_df[toronto_df["Neighborhood"] == ""]
no_neighborhood.head()

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

**Result:** there are no Boroughs without unassigned Neighborhoods.

## Part 2 - adding lattitude and longitude coordinates

Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.

In [11]:
import numpy as np # library to handle data in a vectorized manner
import json # library to handle JSON files
import urllib.request
import urllib.parse
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

**Using OpenCage Geocoding API** to get the longitude and latitude because **Geocoder Python doesn't work**.

https://opencagedata.com/api#forward-opt

Definning the function fetching coordinates based on address
It returns coordingates like [-79.3577, 43.8015] longitude, latitude

In [12]:
# @hidden_cell
my_key='3e3cc12415d54fa598003cc257642df1'

In [13]:
def get_coordinates(postal_code):
    key=my_key
    address = '{}, Toronto, Ontario'.format(postal_code)
    parsed_address = urllib.parse.quote(address)
    download_url = 'https://api.opencagedata.com/geocode/v1/geojson?q={}&key={}'.format(parsed_address, key)
    #print(postal_code, download_url)
    with urllib.request.urlopen(download_url) as urlop:
        data = json.loads(urlop.read().decode())
    
    return data['features'][0]['geometry']['coordinates']

In [14]:
#test the function
get_coordinates("M5G")

[-79.386, 43.6564]

**NOTE** these coordinates are slightly diffferent thant the ones from instructions, this is because I'm using a different data source. I've checked and values are correct - pointing to center of each borough.
eg. for M5G function returns 43.6564, -79.386 vs from instructions 43.657952,-79.38738

In [15]:
long_lat = toronto_df["Postal Code"].apply(get_coordinates)

In [16]:
#check if the size is the same
print(toronto_df.shape, long_lat.shape)
toronto_df["Latitude"] = long_lat.apply(lambda x: x[1])
toronto_df["Longitude"] = long_lat.apply(lambda x: x[0])

(103, 3) (103,)


In [17]:
toronto_df.reset_index(drop=True, inplace=True)
toronto_df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.653482 -79.383935

## Part 3 - Exploring neighborhoods

Once the data is in a structured format, replicate the analysis that we did to the New York City dataset to explore and cluster the neighborhoods in the city of Toronto.

Explore and cluster the neighborhoods in Toronto. You can decide to work with only boroughs that contain the word Toronto and then replicate the same analysis we did to the New York City data. It is up to you.

Just make sure:

1. to add enough Markdown cells to explain what you decided to do and to report any observations you make.
2. to generate maps to visualize your neighborhoods and how they cluster together.

In [18]:
import folium # map rendering library

#### Use OpenCage Geocoding API to get the latitude and longitude of Toronto.

In [19]:
# @hidden_cell
key='3e3cc12415d54fa598003cc257642df1'

In [20]:

address = 'Toronto, Ontario'
parsed_address = urllib.parse.quote(address)
download_url = 'https://api.opencagedata.com/geocode/v1/geojson?q={}&key={}'.format(parsed_address, key)
with urllib.request.urlopen(download_url) as urlop:
    data = json.loads(urlop.read().decode())
toronto_long_lat = data['features'][0]['geometry']['coordinates']
toronto_latitude = toronto_long_lat[1]
toronto_longitude = toronto_long_lat[0]

print('The geograpical coordinate of Toronto are {}, {}.'.format(toronto_latitude, toronto_longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Create a map of Toronto with neighborhoods superimposed on top.

In [40]:
map_newyork = folium.Map(location=[toronto_latitude, toronto_longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_df['Latitude'], toronto_df['Longitude'], toronto_df['Borough'], toronto_df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

Next, utilizing the Foursquare API, explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [22]:
# @hidden_cell
CLIENT_ID = 'WCU4DFKQSFBVCJTZEXZ0KGSXIMZNKB4L2PIKD2WQ0AY2DJ21' # your Foursquare ID
CLIENT_SECRET = 'REPQCFDXZFCKTUSRGOTXWJYE3U02L2UTSABPOJ4F4DXY05CN' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: WCU4DFKQSFBVCJTZEXZ0KGSXIMZNKB4L2PIKD2WQ0AY2DJ21
CLIENT_SECRET:REPQCFDXZFCKTUSRGOTXWJYE3U02L2UTSABPOJ4F4DXY05CN


Create a function to get name, location and categories of venues in a radius of 500 meters from the coordinates of each neighborhood in Toronto

In [23]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        #print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Run the above function on each neighborhood and create a new dataframe called Toronto_venues.

In [24]:
LIMIT = 150 # limit of number of venues returned by Foursquare API

toronto_venues = getNearbyVenues(names=toronto_df['Neighborhood'],
                                   latitudes=toronto_df['Latitude'],
                                   longitudes=toronto_df['Longitude']
                                  )

Let's check the size of the resulting dataframe

In [25]:
print(toronto_venues.shape)
toronto_venues.head()

(2632, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Parkwoods              43.653482              -79.383935   
1    Parkwoods              43.653482              -79.383935   
2    Parkwoods              43.653482              -79.383935   
3    Parkwoods              43.653482              -79.383935   
4    Parkwoods              43.653482              -79.383935   

                      Venue  Venue Latitude  Venue Longitude   Venue Category  
0          Downtown Toronto       43.653232       -79.385296     Neighborhood  
1    Nathan Phillips Square       43.652270       -79.383516            Plaza  
2                    Indigo       43.653515       -79.380696        Bookstore  
3              Chatime 日出茶太       43.655542       -79.384684  Bubble Tea Shop  
4  Textile Museum of Canada       43.654396       -79.386500       Art Museum

Let's check how many venues were returned for each neighborhood

In [26]:
toronto_venues[['Neighborhood', 'Venue']].groupby('Neighborhood').count().sort_values(by=["Venue"], ascending=False)

Venue
Neighborhood                                             
Commerce Court, Victoria Hotel                        100
Richmond, Adelaide, King                              100
Stn A PO Boxes                                        100
Garden District, Ryerson                              100
First Canadian Place, Underground city                100
Toronto Dominion Centre, Design Exchange              100
Harbourfront East, Union Station, Toronto Islands     100
St. James Town                                         91
CN Tower, King and Spadina, Railway Lands, Harb...     82
Church and Wellesley                                   77
Lawrence Manor, Lawrence Heights                       75
Parkwoods                                              73
Rouge Hill, Port Union, Highland Creek                 73
Canada Post Gateway Processing Centre                  73
Malvern, Rouge                                         73
Queen's Park, Ontario Provincial Government            73
Berczy Park                                            70
Central Bay Street                                     60
Willowdale, Willowdale East                            55
Kensington Market, Chinatown, Grange Park              48
Little Portugal, Trinity                               48
Parkdale, Roncesvalles                                 44
St. James Town, Cabbagetown                            44
Don Mills                                              41
Runnymede, Swansea                                     39
The Danforth West, Riverdale                           35
Brockton, Parkdale Village, Exhibition Place           34
India Bazaar, The Beaches West                         33
Guildwood, Morningside, West Hill                      30
Thorncliffe Park                                       29
University of Toronto, Harbord                         29
Downsview                                              28
The Annex, North Midtown, Yorkville                    27
Leaside                                                26
High Park, The Junction South                          25
Regent Park, Harbourfront                              25
Steeles West, L'Amoreaux West                          22
Bedford Park, Lawrence Manor East                      22
Davisville                                             22
West Deane Park, Princess Gardens, Martin Grove...     21
Dufferin, Dovercourt Village                           19
Mimico NW, The Queensway West, South of Bloor, ...     17
Willowdale, Newtonbrook                                17
Kennedy Park, Ionview, East Birchmount Park            16
The Kingsway, Montgomery Road, Old Mill North          15
Clarks Corners, Tam O'Shanter, Sullivan                14
Business reply mail Processing Centre, South Ce...     14
South Steeles, Silverstone, Humbergate, Jamesto...     13
Christie                                               12
Fairview, Henry Farm, Oriole                           12
Glencairn                                              12
Cliffside, Cliffcrest, Scarborough Village West        11
Golden Mile, Clairlea, Oakridge                        10
Westmount                                              10
Parkview Hill, Woodbine Gardens                         9
Caledonia-Fairbanks                                     9
Studio District                                         8
Alderwood, Long Branch                                  8
Woodbine Heights                                        7
Victoria Village                                        7
Moore Park, Summerhill East                             7
Davisville North                                        7
The Beaches                                             6
Summerhill West, Rathnelly, South Hill, Forest ...      6
Islington Avenue, Humber Valley Village                 6
Humewood-Cedarvale                                      6
Humberlea, Emery                                        6
Bathurst Manor, Wilson Heights, Downsview North   

#### Let's find out how many unique categories can be curated from all the returned venues

In [27]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 267 uniques categories.


## Analyze Each Neighborhood

In [28]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   American Restaurant  Aquarium  Arcade  Art Gallery  Art Museum  \
0                    0         0       0            0           0   
1                    0         0       0            0           0   
2                    0         0       0            0           0   
3                    0         0       0            0           0   
4                    0         0       0            0           1   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Dealership  \
0                    0                 0                   0                0   
1                    0                 0                   0                0   
2                    0                 0                   0                0   
3                    0                 0                   0                0   
4                    0                 0                   0                0   

   Auto Garage  BBQ Joint  Baby Store  Badminton Court  Bagel Shop  Bakery  \
0            0          0           0                0           0       0   
1            0          0           0                0           0       0   
2            0          0           0                0           0       0   
3            0          0           0                0           0       0   
4            0          0           0                0           0       0   

   Bank  Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0     0    0               0                 0                 0   
1     0    0               0                 0                 0   
2     0    0               0                 0                 0   
3     0    0               0                 0                 0   
4     0    0               0                 0                 0   

   Basketball Stadium  Beach Bar  Beer Bar  Beer Store  Belgian Restaurant  \
0                   0          0         0           0                   0   
1                   0          0         0           0                   0   
2                   0          0         0           0                   0   
3                   0          0         0           0                   0   
4                   0          0         0           0                   0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          1         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Buffet  Building  Burger Joint  \
0               0        0                0       0         0             0   
1               0        0                0       0         0             0   
2               0        0                0       0         0             0   
3               0        0                1       0         0             0   
4               0        0                0       0         0             0   

   Burrito Place  Bus Line  Bus Station  Bus Stop  Business Service  Butcher  \
0              0         0            0         0                 0        0   
1              0         0            0         0                 0        0   
2              0         0            0         0                 0        0   
3              0         0            0         0                 0        0   
4              0         0            0         0            

In [29]:
toronto_onehot.shape

(2632, 267)

### Group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [30]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                           Agincourt     0.000000   
1                              Alderwood, Long Branch     0.000000   
2     Bathurst Manor, Wilson Heights, Downsview North     0.000000   
3                                     Bayview Village     0.000000   
4                   Bedford Park, Lawrence Manor East     0.000000   
5                                         Berczy Park     0.000000   
6                         Birch Cliff, Cliffside West     0.000000   
7        Brockton, Parkdale Village, Exhibition Place     0.000000   
8   Business reply mail Processing Centre, South C...     0.071429   
9   CN Tower, King and Spadina, Railway Lands, Har...     0.024390   
10                                Caledonia-Fairbanks     0.000000   
11              Canada Post Gateway Processing Centre     0.000000   
12                                          Cedarbrae     0.000000   
13                                 Central Bay Street     0.000000   
14                                           Christie     0.000000   
15                               Church and Wellesley     0.025974   
16            Clarks Corners, Tam O'Shanter, Sullivan     0.000000   
17    Cliffside, Cliffcrest, Scarborough Village West     0.000000   
18                     Commerce Court, Victoria Hotel     0.000000   
19                                         Davisville     0.000000   
20                                   Davisville North     0.000000   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn     0.000000   
22                                          Don Mills     0.000000   
23  Dorset Park, Wexford Heights, Scarborough Town...     0.000000   
24                                          Downsview     0.000000   
25                       Dufferin, Dovercourt Village     0.000000   
26      East Toronto, Broadview North (Old East York)     0.000000   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...     0.000000   
28                       Fairview, Henry Farm, Oriole     0.000000   
29             First Canadian Place, Underground city     0.000000   
30    Forest Hill North & West, Forest Hill Road Park     0.000000   
31                           Garden District, Ryerson     0.000000   
32                                          Glencairn     0.000000   
33                    Golden Mile, Clairlea, Oakridge     0.000000   
34                  Guildwood, Morningside, West Hill     0.000000   
35  Harbourfront East, Union Station, Toronto Islands     0.000000   
36                      High Park, The Junction South     0.000000   
37                                  Hillcrest Village     0.000000   
38                                      Humber Summit     0.000000   
39                                   Humberlea, Emery     0.000000   
40                                 Humewood-Cedarvale     0.000000   
41                     India Bazaar, The Beaches West     0.000000   
42            Islington Avenue, Humber Valley Village     0.000000   
43        Kennedy Park, Ionview, East Birchmount Park     0.000000   
44          Kensington Market, Chinatown, Grange Park     0.000000   
45  Kingsview Village, St. Phillips, Martin Grove ...     0.000000   
46                   Lawrence Manor, Lawrence Heights     0.000000   
47                                      Lawrence Park     0.000000   
48                                            Leaside     0.000000   
49                           Little Portugal, Trinity     0.000000   
50                                     Malvern, Rouge     0.000000   
51  Milliken, Agincourt North, Steeles East, L'Amo...     0.000000   
52  Mimico NW, The Queensway West, South of Bloor,...     0.058824   
53                        Moore Park, Summerhill East     0.000000   
54           North Park, Maple Leaf Park, Upwood Park     0.000000   
55                 North Toronto West,  Lawrence Park     0.000000   
56                Northwest, West Humber - Clairvil

In [31]:
toronto_grouped.shape

(97, 267)

 #### Print each neighborhood along with the top 5 most common venues

In [32]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0             Breakfast Spot  0.25
1            Badminton Court  0.25
2  Latin American Restaurant  0.25
3               Skating Rink  0.25
4               Noodle House  0.00


----Alderwood, Long Branch----
               venue  freq
0        Pizza Place  0.12
1  Convenience Store  0.12
2     Sandwich Place  0.12
3           Pharmacy  0.12
4                Pub  0.12


----Bathurst Manor, Wilson Heights, Downsview North----
                       venue  freq
0              Deli / Bodega  0.17
1                Pizza Place  0.17
2  Middle Eastern Restaurant  0.17
3   Mediterranean Restaurant  0.17
4                Coffee Shop  0.17


----Bayview Village----
                venue  freq
0         Flower Shop   0.2
1  Golf Driving Range   0.2
2               Trail   0.2
3                Park   0.2
4         Gas Station   0.2


----Bedford Park, Lawrence Manor East----
                venue  freq
0         Coffee Shop  0.09
1      Sandwich

                 venue  freq
0         Home Service  0.50
1          Auto Garage  0.25
2  Rental Car Location  0.25
3        Moving Target  0.00
4            Nightclub  0.00


----Humberlea, Emery----
            venue  freq
0     Coffee Shop  0.17
1   Grocery Store  0.17
2  Discount Store  0.17
3       Nightclub  0.17
4            Café  0.17


----Humewood-Cedarvale----
           venue  freq
0   Hockey Arena  0.17
1          Field  0.17
2          Trail  0.17
3  Deli / Bodega  0.17
4  Grocery Store  0.17


----India Bazaar, The Beaches West----
         venue  freq
0          Pub  0.09
1    BBQ Joint  0.06
2  Coffee Shop  0.06
3       Bakery  0.06
4  Pizza Place  0.06


----Islington Avenue, Humber Valley Village----
           venue  freq
0       Pharmacy  0.33
1  Grocery Store  0.17
2           Park  0.17
3   Skating Rink  0.17
4           Bank  0.17


----Kennedy Park, Ionview, East Birchmount Park----
            venue  freq
0  Discount Store  0.19
1     Coffee Shop  0.19
2      

                     venue  freq
0        Convenience Store   0.5
1                     Park   0.5
2              Yoga Studio   0.0
3            Movie Theater   0.0
4  New American Restaurant   0.0


----Wexford, Maryvale----
                     venue  freq
0              Auto Garage   1.0
1              Yoga Studio   0.0
2            Moving Target   0.0
3                Nightclub   0.0
4  New American Restaurant   0.0


----Willowdale, Newtonbrook----
               venue  freq
0  Korean Restaurant  0.24
1        Coffee Shop  0.12
2      Hot Dog Joint  0.06
3              Diner  0.06
4        Pizza Place  0.06


----Willowdale, Willowdale East----
               venue  freq
0  Korean Restaurant  0.16
1    Bubble Tea Shop  0.15
2        Coffee Shop  0.07
3               Café  0.05
4        Pizza Place  0.05


----Willowdale, Willowdale West----
          venue  freq
0  Intersection  0.25
1   Coffee Shop  0.25
2    Baby Store  0.25
3          Park  0.25
4     Nightclub  0.00


----Wobu

#### put that into a pandas dataframe

First, let's write a function to sort the venues in descending order.

In [33]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [34]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood     1st Most Common Venue  \
0                                        Agincourt           Badminton Court   
1                           Alderwood, Long Branch                  Pharmacy   
2  Bathurst Manor, Wilson Heights, Downsview North  Mediterranean Restaurant   
3                                  Bayview Village               Gas Station   
4                Bedford Park, Lawrence Manor East               Coffee Shop   

       2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0  Latin American Restaurant          Skating Rink             Breakfast Spot   
1          Convenience Store    Athletics & Sports                        Pub   
2                Pizza Place   Fried Chicken Joint  Middle Eastern Restaurant   
3                Flower Shop                 Trail                       Park   
4                 Restaurant        Sandwich Place         Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0                 Field  Fast Food Restaurant               Farmers Market   
1           Coffee Shop        Sandwich Place                          Gym   
2         Deli / Bodega           Coffee Shop  Eastern European Restaurant   
3    Golf Driving Range           Event Space         Ethiopian Restaurant   
4         Grocery Store     Indian Restaurant                     Pharmacy   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0    Falafel Restaurant           Event Space   Ethiopian Restaurant  
1           Pizza Place  Ethiopian Restaurant            Event Space  
2      Doner Restaurant            Donut Shop    Dumpling Restaurant  
3     Electronics Store    Falafel Restaurant           Dessert Shop  
4           Pizza Place                   Pub                   Café

## 4. Cluster Neighborhoods

Run k-means to cluster the neighborhood into 5 clusters.

In [35]:
# import k-means from clustering stage
from sklearn.cluster import KMeans
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [37]:
#toronto_merged = toronto_merged_copy.copy()
#neighborhoods_venues_sorted = neighborhoods_venues_sorted_copy.copy()
#toronto_merged_copy = toronto_merged.copy()
#neighborhoods_venues_sorted_copy = neighborhoods_venues_sorted.copy()

In [38]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.653482 -79.383935             1.0        Clothing Store   
1  43.727600 -79.314800             1.0           Pizza Place   
2  43.655500 -79.362600             1.0           Coffee Shop   
3  43.722300 -79.450400             1.0        Clothing Store   
4  43.653482 -79.383935             1.0        Clothing Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Coffee Shop            Restaurant                 Plaza   
1  Portuguese Restaurant     French Restaurant          Intersection   
2             Restaurant        Breakfast Spot                   Pub   
3            Coffee Shop            Restaurant        Cosmetics Shop   
4            Coffee Shop            Restaurant                 Plaza   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0               Theater                 Diner        Cosmetics Shop   
1                  Park           Coffee Shop          Hockey Arena   
2            Food Truck   Distribution Center                Bakery   
3      Sushi Restaurant                Bakery            Food Court   
4               Theater                 Diner        Cosmetics Shop   

  8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
0       Thai Restaurant           Seafood Restaurant    American Restaurant  
1     Electronics Store  Eastern European Restaurant   Ethiopian Restaurant  
2       Thai Restaurant                      Theater     Italian Restaurant  
3     Electronics Store                Women's Store            Men's Store  
4       Thai Restaurant           Seafood Restaurant    American Restaurant

Let's remove boroughs without assigned clusters. And convert cluster labels from float to int

In [41]:
toronto_merged = toronto_merged[np.isnan(toronto_merged["Cluster Labels"])==False]

In [42]:
toronto_merged["Cluster Labels"] = toronto_merged["Cluster Labels"].astype(int)

Finally, let's visualize the resulting clusters

In [43]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[toronto_latitude, toronto_longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

In [44]:
# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examine Clusters
Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

#### Cluster 1

In [45]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
27        North York               0                  Park   
32       Scarborough               0         Grocery Store   
33        North York               0                  Park   
35         East York               0                  Park   
61   Central Toronto               0                  Park   
64              York               0                  Park   
66        North York               0                  Park   
73   Central Toronto               0            Playground   
83   Central Toronto               0                  Park   
91  Downtown Toronto               0           Candy Store   

                       2nd Most Common Venue        3rd Most Common Venue  \
27  Residential Building (Apartment / Condo)  Eastern European Restaurant   
32                                      Park          Xinjiang Restaurant   
33                                  Pharmacy              Thai Restaurant   
35                         Convenience Store                  Coffee Shop   
61                        Photography Studio          Xinjiang Restaurant   
64                         Convenience Store          Xinjiang Restaurant   
66                                      Bank            Convenience Store   
73                                      Park                     Gym Pool   
83                                       Gym              Thai Restaurant   
91                                Playground                Grocery Store   

          4th Most Common Venue 5th Most Common Venue  \
27          Distribution Center               Dog Run   
32  Eastern European Restaurant   Distribution Center   
33            Convenience Store           Coffee Shop   
35          Xinjiang Restaurant   Distribution Center   
61          Distribution Center               Dog Run   
64                      Dog Run      Doner Restaurant   
66            Electronics Store               Dog Run   
73                       Garden                 Diner   
83                Grocery Store          Tennis Court   
91                         Park   Dumpling Restaurant   

        6th Most Common Venue 7th Most Common Venue  \
27           Doner Restaurant            Donut Shop   
32                    Dog Run      Doner Restaurant   
33  Middle Eastern Restaurant           Pizza Place   
35                    Dog Run      Doner Restaurant   
61           Doner Restaurant            Donut Shop   
64                 Donut Shop   Dumpling Restaurant   
66           Doner Restaurant            Donut Shop   
73             Discount Store   Distribution Center   
83                 Playground           Event Space   
91                      Diner        Discount Store   

          8th Most Common Venue        9th Most Common Venue  \
27          Dumpling Restaurant          Xinjiang Restaurant   
32                   Donut Shop          Dumpling Restaurant   
33                    Wine Shop                 Gourmet Shop   
35                   Donut Shop          Dumpling Restaurant   
61          Dumpling Restaurant  Eastern European Restaurant   
64  Eastern European Restaurant            Electronics Store   
66          Dumpling Restaurant  Eastern European Restaurant   
73                      Dog Run             Doner Restaurant   
83         Ethiopian Restaurant            Electronics Store   
91          Distribution Center                      Dog Run   

         10th Most Common Venue  
27                        Diner  
32            Electronics Store  
33                Deli / Bodega  
35  Eastern European Restaurant  
61            Electronics Store  
64         Ethiopian Restaurant  
66          Xinjiang Restaurant  
73                   Donut Shop  
83  Eastern European Restaurant  
91             Doner Restaurant

#### Cluster 2

In [46]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
0          North York               1              Clothing Store   
1          North York               1                 Pizza Place   
2    Downtown Toronto               1                 Coffee Shop   
3          North York               1              Clothing Store   
4    Downtown Toronto               1              Clothing Store   
5           Etobicoke               1                    Pharmacy   
6         Scarborough               1              Clothing Store   
7          North York               1                  Restaurant   
8           East York               1                 Pizza Place   
9    Downtown Toronto               1                 Coffee Shop   
10         North York               1                 Pizza Place   
11          Etobicoke               1                 Coffee Shop   
12        Scarborough               1              Clothing Store   
13         North York               1                  Restaurant   
14          East York               1                       Diner   
15   Downtown Toronto               1                 Coffee Shop   
16               York               1               Deli / Bodega   
18        Scarborough               1                 Pizza Place   
19       East Toronto               1           Health Food Store   
20   Downtown Toronto               1                 Coffee Shop   
21               York               1                        Park   
22        Scarborough               1                 Coffee Shop   
23          East York               1           Electronics Store   
24   Downtown Toronto               1                 Coffee Shop   
25   Downtown Toronto               1                        Café   
26        Scarborough               1  Construction & Landscaping   
28         North York               1    Mediterranean Restaurant   
29          East York               1           Indian Restaurant   
30   Downtown Toronto               1                        Café   
31       West Toronto               1                      Bakery   
34         North York               1              Sandwich Place   
36   Downtown Toronto               1                 Coffee Shop   
37       West Toronto               1                         Bar   
38        Scarborough               1                 Coffee Shop   
39         North York               1                 Gas Station   
40         North York               1                 Pizza Place   
41       East Toronto               1            Greek Restaurant   
42   Downtown Toronto               1                 Coffee Shop   
43       West Toronto               1          Tibetan Restaurant   
44        Scarborough               1                      Bakery   
45         North York               1           Martial Arts Dojo   
46         North York               1                 Pizza Place   
47       East Toronto               1                         Pub   
48   Downtown Toronto               1                 Coffee Shop   
49         North York               1                      Bakery   
51        Scarborough               1              Ice Cream Shop   
52         North York               1           Korean Restaurant   
53         North York               1                 Pizza Place   
54       East Toronto               1               Garden Center   
55         North York               1                 Coffee Shop   
56               York               1        Fast Food Restaurant   
57         North York               1                 Coffee Shop   
58        Scarborough               1                Skating Rink   
59         North York               1           Korean Restaurant   
60         North York               1                 Pizza Place   
63               York               1                     Brewery   
65        Scarborough               1                      Bakery   
67    Centra

#### Cluster 3

In [47]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
62  Central Toronto               2           IT Services   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
62     Electronics Store   Distribution Center               Dog Run   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
62      Doner Restaurant            Donut Shop   Dumpling Restaurant   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
62  Eastern European Restaurant   Xinjiang Restaurant                  Diner

#### Cluster 4

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

#### Cluster 5

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]